README.md
# Python Hand Tracking Module

This is a Hand Tracking Module, which tracks the position of different landmarks in the hand.

------------------------------------------------------------------
Package at Pypi : https://pypi.org/project/Silver-HandTrackingModule/

How to install : `pip install Silver-HandTrackingModule`

------------------------------------------------------------------

1. Language Used : **Python**
2. Requirements(Do a pip install) : 
   - OpenCV
   - MediaPipe
3. Files :
   - **activate.bat** : Just activte the virtual enviroment, like a shortcut.
   - **HandTrackingModule.py** - Main Module. You Need to import this file to your File.
   - **media pipe hand landmarks.PNG** - Hand Land marks Used By the Mediapipe.
   - **main.py** - The Whole Thing. I used it to devolop the project.
   - **Application2.py** - Used this to show how to use the Hand Tracking Module.
   - **README.md** : Summon Michael Jackson.

My **My pip list**
```
(venv) D:\Projects\Python\HandTrack>pip list
Package               Version
--------------------- --------
absl-py               0.13.0  
attrs                 21.2.0  
cycler                0.10.0  
kiwisolver            1.3.2   
matplotlib            3.4.3   
mediapipe             0.8.7.1 
numpy                 1.21.2  
opencv-contrib-python 4.5.3.56
opencv-python         4.5.3.56
Pillow                8.3.1   
pip                   21.1.3  
protobuf              3.17.3  
pyparsing             2.4.7   
python-dateutil       2.8.2   
setuptools            56.0.0
six                   1.16.0
wheel                 0.37.0
```


HandTrackingModule/CHANGELOG.txt
Change Log
==========

1.0.0 (17-09-2021)
-------------------
- First Relase

main.py
import cv2
import mediapipe as mp
import time

# camera selector
cap = cv2.VideoCapture(0)

mpHands = mp.solutions.hands
hands = mpHands.Hands()
mpDraw = mp.solutions.drawing_utils

#FPS
pTime = 0
cTime = 0

while True:
    success, img = cap.read()
    imgRGB = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    results = hands.process(imgRGB)

    if results.multi_hand_landmarks:
        for handLms in results.multi_hand_landmarks:
            for id, lm in enumerate(handLms.landmark):
                # print(id,lm)
                h, w, c = img.shape
                cx, cy = int(lm.x * w), int(lm.y * h)
                print(id, cx, cy)
                #Pin point a landmark
                if id == 4:
                    cv2.circle(img, (cx,cy), 15, (255,0,255), cv2.FILLED)

            mpDraw.draw_landmarks(img, handLms, mpHands.HAND_CONNECTIONS)

    cTime = time.time()
    fps = 1/(cTime-pTime)
    pTime = cTime

    cv2.putText(img, str(int(fps)), (10,70), cv2.FONT_HERSHEY_COMPLEX, 2, (255,0,255), 2)

    cv2.imshow("Image", img)
    cv2.waitKey(1)

HandTrackingModule.py
import cv2
import mediapipe as mp
import time

class handDetector():
    def __init__(self, mode=False, maxHands=2, detectionCon=0.5, trackCon=0.5):
        self.mode = mode
        self.maxHands = maxHands
        self.detectionCon = detectionCon
        self.trackCon = trackCon

        self.mpHands = mp.solutions.hands
        self.hands = self.mpHands.Hands(self.mode, self.maxHands, self.detectionCon, self.trackCon)
        self.mpDraw = mp.solutions.drawing_utils

    def findHands(self,img, draw=True):
        imgRGB = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
        self.results = self.hands.process(imgRGB)

        if self.results.multi_hand_landmarks:
            for handLms in self.results.multi_hand_landmarks:
                if draw:
                    self.mpDraw.draw_landmarks(img, handLms, self.mpHands.HAND_CONNECTIONS)
        return img

    def findPosition(self, img, handNo=0, draw=True):
        lmList = []
        if self.results.multi_hand_landmarks:
            myHand = self.results.multi_hand_landmarks[handNo]
            for id, lm in enumerate(myHand.landmark):
                h, w, c = img.shape
                cx, cy = int(lm.x * w), int(lm.y * h)
                lmList.append([id, cx, cy])
                #Draw point a landmark
                if draw:
                    cv2.circle(img, (cx,cy), 8, (255,0,255), cv2.FILLED)
        return lmList

def main():
    #FPS
    pTime = 0
    cTime = 0

    # camera selector
    cap = cv2.VideoCapture(0)
    detector = handDetector()

    while True:
        success, img = cap.read()
        img = detector.findHands(img)
        lmList = detector.findPosition(img)        

        if(len(lmList) != 0):
            print(lmList[4])

        cTime = time.time()
        fps = 1/(cTime-pTime)
        pTime = cTime

        cv2.putText(img, str(int(fps)), (10,70), cv2.FONT_HERSHEY_COMPLEX, 2, (255,0,255), 2)

        cv2.imshow("Image", img)
        cv2.waitKey(1)

# print(__name__)
if __name__ == "__main__":
    print("__name__")
    main()

HandTrackingModule/Silver-HandTrackingModule/__init__.py
import cv2
import mediapipe as mp
import time

class handDetector():
    def __init__(self, mode=False, maxHands=2, detectionCon=0.5, trackCon=0.5):
        self.mode = mode
        self.maxHands = maxHands
        self.detectionCon = detectionCon
        self.trackCon = trackCon

        self.mpHands = mp.solutions.hands
        self.hands = self.mpHands.Hands(self.mode, self.maxHands, self.detectionCon, self.trackCon)
        self.mpDraw = mp.solutions.drawing_utils

    def findHands(self,img, draw=True):
        imgRGB = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
        self.results = self.hands.process(imgRGB)

        if self.results.multi_hand_landmarks:
            for handLms in self.results.multi_hand_landmarks:
                if draw:
                    self.mpDraw.draw_landmarks(img, handLms, self.mpHands.HAND_CONNECTIONS)
        return img

    def findPosition(self, img, handNo=0, draw=True):
        lmList = []
        if self.results.multi_hand_landmarks:
            myHand = self.results.multi_hand_landmarks[handNo]
            for id, lm in enumerate(myHand.landmark):
                h, w, c = img.shape
                cx, cy = int(lm.x * w), int(lm.y * h)
                lmList.append([id, cx, cy])
                #Draw point a landmark
                if draw:
                    cv2.circle(img, (cx,cy), 10, (255,0,255), cv2.FILLED)
        return lmList

def main():
    #FPS
    pTime = 0
    cTime = 0
    # camera selector
    cap = cv2.VideoCapture(0)
    detector = handDetector()

    while True:
        success, img = cap.read()
        img = detector.findHands(img)
        lmList = detector.findPosition(img)        

        if(len(lmList) != 0):
            print(lmList[4])

        cTime = time.time()
        fps = 1/(cTime-pTime)
        pTime = cTime

        cv2.putText(img, str(int(fps)), (10,70), cv2.FONT_HERSHEY_COMPLEX, 2, (255,0,255), 2)

        cv2.imshow("Image", img)
        cv2.waitKey(1)

# print(__name__)
if __name__ == "__main__":
    print("__name__")
    main()